internal/config/config.go

```diff
@@ -0,0 +1,161 @@
+package config
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+)
+
+// validateDistributions checks that the condition that
+// two distributions with the same distribution ID MUST NOT share the same prefix.
+// or in other terms, every pair of id,prefix (Distribution) must be unique
+func validateDistributions(distributions distributionsMap) error {
+	uniqueMap := make(map[string]struct{})
+
+	for _, value := range distributions {
+		hash := value.stringPropertiesHash()
+		if _, ok := uniqueMap[hash]; ok {
+			return fmt.Errorf("error parsing configuration: distribution value duplicated id: %s prefix: %s", value.ID, value.Prefix)
+		}
+
+		uniqueMap[hash] = struct{}{}
+	}
+
+	return nil
+}
+
+func (c *Config) parse(data []byte) error {
+	config := struct {
+		Distributions distributionsMap `json:"distributions"`
+		Entitlements  entitlementsMap  `json:"entitlements"`
+	}{}
+
+	if err := yaml.Unmarshal(data, config); err != nil {
+		return err
+	}
+
+	err := validateDistributions(config.Distributions)
+	if err != nil {
+		return err
+	}
+
+	for name, value := range config.Distributions {
+		c.distributions.Set(name, value)
+	}
+	for _, name := range c.distributions.Names() {
+		if _, ok := config.Distributions[name]; !ok {
+			c.distributions.Delete(name)
+		}
+	}
+
+	for name, value := range config.Entitlements {
+		c.entitlements.Set(name, value)
+	}
+	for _, name := range c.entitlements.Names() {
+		if _, ok := config.Entitlements[name]; !ok {
+			c.entitlements.Delete(name)
+		}
+	}
```

Comment:
This logic is not covered in the tests.

Reply:
yes that's the idea, parse can be called concurrently, as I understand this is the way it should be done but maybe there's a better/easier way

Reply:
I just tried a simple program where the map gets cleared but a variable that held a previous pointer still works https://go.dev/play/p/ZtCrUHoyueE. So I think maybe this code block can be simplified from Set/Delete/Names with a mutex protecting each, to just simply lock mutex, create new map, populate it, and unlock mutex.

But I may be wrong or missing something, and not super opinionated on how it's done.